File: Mage/StaticPointAnnotation.m
//
//  StaticPointAnnotation.m
//  MAGE
//
//

#import "StaticPointAnnotation.h"

@implementation StaticPointAnnotation

-(id) initWithFeature:(NSDictionary *)feature {
    if ((self = [super init])) {
        _feature = feature;
        // Set a title so that the annotation tap event will actually occur on the map delegate
        self.title = @" ";
        NSArray *coordinates = [_feature valueForKeyPath:@"geometry.coordinates"];
        [self setCoordinate:CLLocationCoordinate2DMake([[coordinates objectAtIndex: 1] floatValue], [[coordinates objectAtIndex: 0] floatValue])];
        _iconUrl = [_feature valueForKeyPath:@"properties.style.iconStyle.icon.href"];
    }
    return self;
}

- (MKAnnotationView *) viewForAnnotationOnMapView: (MKMapView *) mapView {
    if (self.iconUrl == nil) {
        return [self defaultAnnotationView:mapView];
    } else {
        return [self customAnnotationView:mapView];
    }
}

- (MKAnnotationView *) defaultAnnotationView: (MKMapView *) mapView {
    MKAnnotationView *annotationView = (MKAnnotationView *) [mapView dequeueReusableAnnotationViewWithIdentifier:@"pinAnnotation"];
    if (annotationView == nil) {
        annotationView = [[MKPinAnnotationView alloc] initWithAnnotation:self reuseIdentifier:@"pinAnnotation"];
    } else {
        annotationView.annotation = self;
    }
    
    return annotationView;
}

- (MKAnnotationView *) customAnnotationView: (MKMapView *) mapView {
    MKAnnotationView *annotationView = (MKAnnotationView *) [mapView dequeueReusableAnnotationViewWithIdentifier:_iconUrl];
    
    if (annotationView == nil) {
        NSLog(@"showing icon from %@", self.iconUrl);
        
        annotationView = [[MKAnnotationView alloc] initWithAnnotation:self reuseIdentifier:self.iconUrl];
        annotationView.enabled = YES;
        annotationView.canShowCallout = YES;
        UIImage *image = nil;
        if ([[self.iconUrl lowercaseString] hasPrefix:@"http"]) {
            image = [UIImage imageWithData:[NSData dataWithContentsOfURL:[NSURL URLWithString:self.iconUrl]]];
        } else {
            NSString *documentsDirectory = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)objectAtIndex:0];
            image = [UIImage imageWithData:[NSData dataWithContentsOfFile:[NSString stringWithFormat:@"%@/%@", documentsDirectory,self.iconUrl]]];
        }
        
        CGFloat scale = image.size.width / 35.0;
        annotationView.image = [UIImage imageWithCGImage:[image CGImage] scale:scale orientation:image.imageOrientation];
        
        annotationView.centerOffset = CGPointMake(0, -(annotationView.image.size.height/2.0f));
        annotationView.annotation = self;
    } else {
        annotationView.annotation = self;
    }
    
    return annotationView;
}

- (UIView *) detailViewForAnnotation {
    UIView *view = [[UIView alloc] init];
    view.translatesAutoresizingMaskIntoConstraints = false;
    [view addConstraint:[NSLayoutConstraint constraintWithItem:view attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationLessThanOrEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:300]];
    [view addConstraint:[NSLayoutConstraint constraintWithItem:view attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:200]];
    
    NSString *description = [_feature valueForKeyPath: @"properties.description"];
    
    NSMutableAttributedString *attributedDescription = [[NSMutableAttributedString alloc] initWithData:[description dataUsingEncoding:NSUTF8StringEncoding]
                                                                                 options:@{NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType,
                                                                                           NSCharacterEncodingDocumentAttribute: @(NSUTF8StringEncoding)}
                                                                      documentAttributes:nil error:nil];
    NSString *titleText = [_feature valueForKeyPath: @"properties.name"];
    self.title = nil;
    UILabel *titleLabel = [[UILabel alloc] init];
    titleLabel.text = titleText;
    titleLabel.font = [UIFont systemFontOfSize:20];
    titleLabel.translatesAutoresizingMaskIntoConstraints = false;
    titleLabel.numberOfLines = 1;
    [view addSubview:titleLabel];

    UILabel *descriptionLabel = [[UILabel alloc] init];
    descriptionLabel.attributedText = attributedDescription;
    descriptionLabel.font = [UIFont systemFontOfSize:12];
    descriptionLabel.translatesAutoresizingMaskIntoConstraints = false;
    descriptionLabel.numberOfLines = 0;
    
    CGFloat labelHeight = [self getLabelHeight:descriptionLabel];
    
    [descriptionLabel addConstraint:[NSLayoutConstraint constraintWithItem:descriptionLabel attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:labelHeight]];
    [descriptionLabel addConstraint:[NSLayoutConstraint constraintWithItem:descriptionLabel attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:200]];
    [view addConstraint:[NSLayoutConstraint constraintWithItem:view attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1 constant:labelHeight]];
    
    [view addSubview:descriptionLabel];

    NSDictionary *views = NSDictionaryOfVariableBindings(titleLabel, descriptionLabel);

    [view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[titleLabel]|" options:0 metrics:nil views:views]];
    [view addConstraint:[NSLayoutConstraint constraintWithItem:descriptionLabel attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:view attribute:NSLayoutAttributeLeft multiplier:1 constant:0]];
    [view addConstraint:[NSLayoutConstraint constraintWithItem:descriptionLabel attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:view attribute:NSLayoutAttributeTop multiplier:1 constant:titleText == nil ? 0 :30]];
    [view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[titleLabel]-[descriptionLabel]|" options:0 metrics:nil views:views]];

    return view;
}

- (CGFloat)getLabelHeight:(UILabel*)label
{
    CGSize constraint = CGSizeMake(label.frame.size.width, CGFLOAT_MAX);
    CGSize size;
    
    NSStringDrawingContext *context = [[NSStringDrawingContext alloc] init];
    CGSize boundingBox = [label.attributedText boundingRectWithSize:constraint options:NSStringDrawingUsesLineFragmentOrigin context:context].size;
    
    size = CGSizeMake(ceil(boundingBox.width), ceil(boundingBox.height));
    
    return size.height;
}

@end

File: Mage/StaticPointAnnotation.swift
//
//  StaticPointAnnotation.swift
//  MAGE
//
//
import MapKit
import MapFramework
import DataSourceDefinition

class StaticPointAnnotation: DataSourceAnnotation {
    override var dataSource: any DataSourceDefinition {
        get {
            DataSources.featureItem
        }
        set { }
    }
//    var feature: [AnyHashable: Any]?
    var iconUrl: String?
    var layerName: String?
    var title: String?
    var subtitle: String?
    var view: MKAnnotationView?
    var timestamp: Date?
    
    public init(feature: [AnyHashable: Any]) {
//        self.feature = feature
        self.title = {
            if let properties = feature["properties"] as? [AnyHashable: Any],
               let name = properties["name"] as? String
            {
                return name
            }
            return " "
        }()
        
        self.subtitle = {
            if let properties = feature["properties"] as? [AnyHashable: Any],
               let description = properties["description"] as? String
            {
                return description
            }
            return nil
        }()
        
        self.timestamp = {
            guard let timestamp = (feature["properties"] as? [AnyHashable : Any])?["timestamp"] as? String else {
                return nil
            }
            let lastModifiedDate = ISO8601DateFormatter.gmtZeroDate(from: timestamp) ?? Date();
            return lastModifiedDate
        }()
        
        var coordinate: CLLocationCoordinate2D = kCLLocationCoordinate2DInvalid
        if let coordinates = (feature["geometry"] as? [AnyHashable: Any])?["coordinates"] as? [Double] {
            coordinate = CLLocationCoordinate2D(latitude: coordinates[1], longitude: coordinates[0])
        }
        if let properties = feature["properties"] as? [AnyHashable: Any],
           let style = properties["style"] as? [AnyHashable: Any],
           let iconStyle = style["iconStyle"] as? [AnyHashable: Any],
           let icon = iconStyle["icon"] as? [AnyHashable: Any],
           let href = icon["href"] as? String {
            self.iconUrl = href
        }
        var iconURL: URL?
        if let iconUrlStr = self.iconUrl {
            if iconUrlStr.hasPrefix("http") {
                iconURL = URL(string: iconUrlStr)
            } else {
                iconURL = URL(fileURLWithPath: "\(FeatureItem.getDocumentsDirectory())/\(iconUrlStr)")
            }
        }
        let fi = FeatureItem(
            featureDetail: StaticLayer.featureDescription(feature: feature),
            coordinate: coordinate,
            featureTitle: StaticLayer.featureName(feature: feature),
            layerName: "",
            iconURL: iconURL,
            featureDate: StaticLayer.featureTimestamp(feature: feature)
        )
        super.init(coordinate: coordinate, itemKey: fi.toKey())
    }
    
    func viewForAnnotation(on mapView: MKMapView, scheme: MDCContainerScheming?) -> MKAnnotationView {
        if let iconUrl = self.iconUrl {
            return customAnnotationView(mapView: mapView, iconUrl: iconUrl)
        }
        return defaultAnnotationView(mapView: mapView)
    }
    
    func defaultAnnotationView(mapView: MKMapView) -> MKAnnotationView {
        let annotationView: MKAnnotationView = {
            if let annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: "pinAnnotation") {
                annotationView.annotation = self
                return annotationView
            }
            let annotationView = MKMarkerAnnotationView(annotation: self, reuseIdentifier: "pinAnnotation")
            annotationView.titleVisibility = .hidden
            annotationView.subtitleVisibility = .hidden
            annotationView.canShowCallout = false
            annotationView.isEnabled = false
            return annotationView
        }()
        
        self.view = annotationView
        return annotationView
    }
    
    func customAnnotationView(mapView: MKMapView, iconUrl: String) -> MKAnnotationView {
        if let annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: iconUrl) {
            annotationView.annotation = self
            self.view = annotationView
            return annotationView
        }
        NSLog("Showing icon from \(iconUrl)")
        let annotationView = MKAnnotationView(annotation: self, reuseIdentifier: iconUrl)
        annotationView.isEnabled = false
        annotationView.canShowCallout = false
        DispatchQueue.global(qos: .userInitiated).async {
            let image: UIImage? = {
                if iconUrl.lowercased().hasPrefix("http"), let iconUrl = URL(string: iconUrl) {
                    if let data = try? Data(contentsOf: iconUrl) {
                        return UIImage(data: data)
                    }
                } else {
                    if let data = try? Data(contentsOf: self.getDocumentsDirectory().appending(component: iconUrl)) {
                        return UIImage(data: data)
                    }
                }
                return UIImage(named: "marker")
            }()
            if let image = image {
                let widthScale = 35.0
                let scaledImage = image.aspectResize(to: CGSize(width: widthScale, height: image.size.height / (image.size.width / widthScale)))
                Task { @MainActor in
                    annotationView.image = scaledImage
                    annotationView.centerOffset = CGPoint(x: 0, y: -(scaledImage.size.height / 2.0))
                }
            }
        }
        annotationView.annotation = self
        self.view = annotationView
        return annotationView
    }
    
    func getDocumentsDirectory() -> URL {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        let documentsDirectory = paths[0]
        return documentsDirectory
    }
    
    func detailTextForAnnotation() -> String {
        return "\(title ?? "")</br>\(subtitle ?? "")"
    }
}
